File: android/app/src/main/java/com/openclaw/console/data/model/Deployment.kt
package com.openclaw.console.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverter
import androidx.room.TypeConverters
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.time.Instant
import java.util.*

// MARK: - Deployment Status

enum class DeploymentStatus(val value: String) {
    PENDING("pending"),
    RUNNING("running"),
    COMPLETED("completed"),
    FAILED("failed"),
    CANCELLED("cancelled");

    val displayName: String
        get() = when (this) {
            PENDING -> "Pending"
            RUNNING -> "Running"
            COMPLETED -> "Completed"
            FAILED -> "Failed"
            CANCELLED -> "Cancelled"
        }

    val isComplete: Boolean
        get() = when (this) {
            COMPLETED, FAILED, CANCELLED -> true
            PENDING, RUNNING -> false
        }

    companion object {
        fun fromValue(value: String): DeploymentStatus =
            entries.find { it.value == value } ?: PENDING
    }
}

// MARK: - Deployment Environment

enum class DeploymentEnvironment(val value: String) {
    STAGING("staging"),
    PRODUCTION("production");

    val displayName: String
        get() = when (this) {
            STAGING -> "Staging"
            PRODUCTION -> "Production"
        }

    val requiresApproval: Boolean
        get() = when (this) {
            STAGING -> false
            PRODUCTION -> true
        }

    companion object {
        fun fromValue(value: String): DeploymentEnvironment =
            entries.find { it.value == value } ?: STAGING
    }
}

// MARK: - Deployment Platform

enum class DeploymentPlatform(val value: String) {
    IOS("ios"),
    ANDROID("android"),
    BOTH("both");

    val displayName: String
        get() = when (this) {
            IOS -> "iOS"
            ANDROID -> "Android"
            BOTH -> "iOS & Android"
        }

    companion object {
        fun fromValue(value: String): DeploymentPlatform =
            entries.find { it.value == value } ?: BOTH
    }
}

// MARK: - Deployment Artifact

data class DeploymentArtifact(
    val id: String,
    val name: String,
    val platform: DeploymentPlatform,
    val version: String,
    val buildNumber: String,
    val size: Long,
    val downloadURL: String?,
    val createdAt: Instant
) {
    fun getFormattedSize(): String {
        return android.text.format.Formatter.formatFileSize(
            // Context would be needed here - for now return raw bytes
            null, size
        ) ?: "${size} bytes"
    }
}

// MARK: - Deployment Step

data class DeploymentStep(
    val id: String,
    val deploymentId: String,
    val name: String,
    val description: String,
    val status: DeploymentStatus,
    val startedAt: Instant?,
    val completedAt: Instant?,
    val error: String?,
    val logs: List<String>
) {
    val duration: Long?
        get() = if (startedAt != null && completedAt != null) {
            completedAt.toEpochMilli() - startedAt.toEpochMilli()
        } else null
}

// MARK: - Deployment

@Entity(tableName = "deployments")
@TypeConverters(DeploymentTypeConverters::class)
data class Deployment(
    @PrimaryKey val id: String,
    val agentId: String,
    val title: String,
    val description: String,
    val environment: DeploymentEnvironment,
    val platform: DeploymentPlatform,
    val status: DeploymentStatus,
    val branch: String,
    val commit: String,
    val commitMessage: String,
    val triggeredBy: String,
    val createdAt: Instant,
    val startedAt: Instant?,
    val completedAt: Instant?,
    val steps: List<DeploymentStep>,
    val artifacts: List<DeploymentArtifact>,
    val approvalId: String?
) {
    val duration: Long?
        get() = if (startedAt != null && completedAt != null) {
            completedAt.toEpochMilli() - startedAt.toEpochMilli()
        } else null

    val shortCommit: String
        get() = commit.take(8)

    val requiresApproval: Boolean
        get() = environment.requiresApproval

    val canCancel: Boolean
        get() = status == DeploymentStatus.PENDING || status == DeploymentStatus.RUNNING
}

// MARK: - Deployment Request

data class DeploymentRequest(
    val environment: DeploymentEnvironment,
    val platform: DeploymentPlatform,
    val branch: String,
    val description: String?
)

// MARK: - WebSocket Payloads

data class DeploymentUpdate(
    val id: String,
    val agentId: String,
    val status: DeploymentStatus,
    val updatedAt: Instant
)

data class DeploymentStepUpdate(
    val id: String,
    val deploymentId: String,
    val status: DeploymentStatus,
    val startedAt: Instant?,
    val completedAt: Instant?,
    val error: String?,
    val logs: List<String>
)

// MARK: - Type Converters for Room

class DeploymentTypeConverters {
    private val gson = Gson()

    @TypeConverter
    fun fromDeploymentStatus(status: DeploymentStatus): String = status.value

    @TypeConverter
    fun toDeploymentStatus(value: String): DeploymentStatus = DeploymentStatus.fromValue(value)

    @TypeConverter
    fun fromDeploymentEnvironment(env: DeploymentEnvironment): String = env.value

    @TypeConverter
    fun toDeploymentEnvironment(value: String): DeploymentEnvironment =
        DeploymentEnvironment.fromValue(value)

    @TypeConverter
    fun fromDeploymentPlatform(platform: DeploymentPlatform): String = platform.value

    @TypeConverter
    fun toDeploymentPlatform(value: String): DeploymentPlatform =
        DeploymentPlatform.fromValue(value)

    @TypeConverter
    fun fromInstant(instant: Instant?): Long? = instant?.toEpochMilli()

    @TypeConverter
    fun toInstant(timestamp: Long?): Instant? = timestamp?.let { Instant.ofEpochMilli(it) }

    @TypeConverter
    fun fromDeploymentStepList(steps: List<DeploymentStep>): String = gson.toJson(steps)

    @TypeConverter
    fun toDeploymentStepList(json: String): List<DeploymentStep> {
        val type = object : TypeToken<List<DeploymentStep>>() {}.type
        return gson.fromJson(json, type) ?: emptyList()
    }

    @TypeConverter
    fun fromDeploymentArtifactList(artifacts: List<DeploymentArtifact>): String =
        gson.toJson(artifacts)

    @TypeConverter
    fun toDeploymentArtifactList(json: String): List<DeploymentArtifact> {
        val type = object : TypeToken<List<DeploymentArtifact>>() {}.type
        return gson.fromJson(json, type) ?: emptyList()
    }

    @TypeConverter
    fun fromStringList(list: List<String>): String = gson.toJson(list)

    @TypeConverter
    fun toStringList(json: String): List<String> {
        val type = object : TypeToken<List<String>>() {}.type
        return gson.fromJson(json, type) ?: emptyList()
    }
}

File: ios/OpenClawConsole/OpenClawConsole/Models/Deployment.swift
// Models/Deployment.swift
// OpenClaw Work Console
// Deployment models for IDP self-service deployment functionality

import Foundation

// MARK: - Deployment Status

enum DeploymentStatus: String, Codable, CaseIterable {
    case pending
    case running
    case completed
    case failed
    case cancelled

    var displayName: String {
        switch self {
        case .pending: return "Pending"
        case .running: return "Running"
        case .completed: return "Completed"
        case .failed: return "Failed"
        case .cancelled: return "Cancelled"
        }
    }

    var systemImage: String {
        switch self {
        case .pending: return "clock"
        case .running: return "gearshape.arrow.triangle.2.circlepath"
        case .completed: return "checkmark.circle.fill"
        case .failed: return "exclamationmark.triangle.fill"
        case .cancelled: return "xmark.circle.fill"
        }
    }

    var isComplete: Bool {
        switch self {
        case .completed, .failed, .cancelled: return true
        case .pending, .running: return false
        }
    }
}

// MARK: - Deployment Environment

enum DeploymentEnvironment: String, Codable, CaseIterable {
    case staging
    case production

    var displayName: String {
        switch self {
        case .staging: return "Staging"
        case .production: return "Production"
        }
    }

    var requiresApproval: Bool {
        switch self {
        case .staging: return false
        case .production: return true
        }
    }

    var systemImage: String {
        switch self {
        case .staging: return "testtube.2"
        case .production: return "globe"
        }
    }
}

// MARK: - Deployment Platform

enum DeploymentPlatform: String, Codable, CaseIterable {
    case ios
    case android
    case both

    var displayName: String {
        switch self {
        case .ios: return "iOS"
        case .android: return "Android"
        case .both: return "iOS & Android"
        }
    }

    var systemImage: String {
        switch self {
        case .ios: return "applelogo"
        case .android: return "a.square"
        case .both: return "square.grid.2x2"
        }
    }
}

// MARK: - Deployment Artifact

struct DeploymentArtifact: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let platform: DeploymentPlatform
    let version: String
    let buildNumber: String
    let size: Int64
    let downloadURL: String?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case platform
        case version
        case buildNumber = "build_number"
        case size
        case downloadURL = "download_url"
        case createdAt = "created_at"
    }

    var formattedSize: String {
        ByteCountFormatter.string(fromByteCount: size, countStyle: .file)
    }

    static func == (lhs: DeploymentArtifact, rhs: DeploymentArtifact) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

// MARK: - Deployment Step

struct DeploymentStep: Codable, Identifiable, Hashable {
    let id: String
    let deploymentId: String
    let name: String
    let description: String
    let status: DeploymentStatus
    let startedAt: Date?
    let completedAt: Date?
    let error: String?
    let logs: [String]

    enum CodingKeys: String, CodingKey {
        case id
        case deploymentId = "deployment_id"
        case name
        case description
        case status
        case startedAt = "started_at"
        case completedAt = "completed_at"
        case error
        case logs
    }

    var duration: TimeInterval? {
        guard let startedAt, let completedAt else { return nil }
        return completedAt.timeIntervalSince(startedAt)
    }

    static func == (lhs: DeploymentStep, rhs: DeploymentStep) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

// MARK: - Deployment

struct Deployment: Codable, Identifiable, Hashable {
    let id: String
    let agentId: String
    let title: String
    let description: String
    let environment: DeploymentEnvironment
    let platform: DeploymentPlatform
    let status: DeploymentStatus
    let branch: String
    let commit: String
    let commitMessage: String
    let triggeredBy: String
    let createdAt: Date
    let startedAt: Date?
    let completedAt: Date?
    let steps: [DeploymentStep]
    let artifacts: [DeploymentArtifact]
    let approvalId: String?

    enum CodingKeys: String, CodingKey {
        case id
        case agentId = "agent_id"
        case title
        case description
        case environment
        case platform
        case status
        case branch
        case commit
        case commitMessage = "commit_message"
        case triggeredBy = "triggered_by"
        case createdAt = "created_at"
        case startedAt = "started_at"
        case completedAt = "completed_at"
        case steps
        case artifacts
        case approvalId = "approval_id"
    }

    var duration: TimeInterval? {
        guard let startedAt, let completedAt else { return nil }
        return completedAt.timeIntervalSince(startedAt)
    }

    var shortCommit: String {
        String(commit.prefix(8))
    }

    var requiresApproval: Bool {
        environment.requiresApproval
    }

    var canCancel: Bool {
        status == .pending || status == .running
    }

    static func == (lhs: Deployment, rhs: Deployment) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

// MARK: - Deployment Request

struct DeploymentRequest: Codable {
    let environment: DeploymentEnvironment
    let platform: DeploymentPlatform
    let branch: String
    let description: String?

    enum CodingKeys: String, CodingKey {
        case environment
        case platform
        case branch
        case description
    }
}

// MARK: - WebSocket Payloads

struct DeploymentUpdate: Codable {
    let id: String
    let agentId: String
    let status: DeploymentStatus
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case agentId = "agent_id"
        case status
        case updatedAt = "updated_at"
    }
}

struct DeploymentStepUpdate: Codable {
    let id: String
    let deploymentId: String
    let status: DeploymentStatus
    let startedAt: Date?
    let completedAt: Date?
    let error: String?
    let logs: [String]

    enum CodingKeys: String, CodingKey {
        case id
        case deploymentId = "deployment_id"
        case status
        case startedAt = "started_at"
        case completedAt = "completed_at"
        case error
        case logs
    }
}